### DISCUSSION TITLE
Seperate getting params and props in getStaticPaths()

### DISCUSSION THREAD

**Discussion by OliverSpeir** [withastro/roadmap#900] [Proposal]
### Body

*Posting on behalf of `@ardesy` on discord*

In `getStaticPaths` it's impossible to know which page had an error generating its props because all params and props are being generated together. It could be generating lots of objects like `{ params: ..., props: ...}` and if theres an error in one of them the whole operation fails.

There's lots of problems with this. In dev mode if one page's props fail all tabs in the browser show the same error even if they loaded different pages. If any of the pages fails none of them can be accessed in dev which is a big problem because it forces you to fix something right now even if you're working on something else. And the console output can't tell you which page's props caused the error because the whole thing failed so you have to go through every possible page generated in `getStaticPaths` to find out which one it is. Things go wrong in development all the time so there needs to be a way to tell you which page's props are wrong.

`getStaticPaths` is a limiting API. It should support a new API returning an object with `getParams` which returns an array of all page params and `getProps` which is called by astro to turn every params into props. If `getProps` has an error astro can tell you exactly `failure in getProps with params: {page:1}` and you can fix it immediately:

```js
export async function getStaticPaths() {
  return {

    // returns array of params like: [ {params:{page:1}}, {params:{page:2}} ]
    getParams: await getParamsFromDB(),

    // takes {page:1} as argument
    getProps: paramsObject => await getPropsFromDB(paramsObject)
  }
}
```

There's lot of benefits: astro knows exactly where the error happened, it can be done for millions of files without running out of memory (with the current API everything has to be loaded in memory), and if there's an error in one page astro can still generate all the other pages.

It wouldn't even be a breaking change because astro would know the difference because `getStaticPaths` is returning an object instead of array. Astro can behave exactly how the user wants depending on whether they return an array or object. And you can combine both in the same project if you really want. Return an array for very simple cases and where everything is hardcoded and return an object for complicated logic and async calls.

**Comment by Fryuni** [comment]
What would be the advantage of that API over this current solution?
```ts
export function getStaticPaths() {
  // returns array of params like: [ {params:{page:1}}, {params:{page:2}} ]
  return getParamsFromDB()
}

// takes {page:1} as argument
const props = await getPropsFromDB(Astro.params);
```

**Reply by Princesseuh** [reply]
I guess if it was directly in `getStaticPaths` you benefits from being able to use a automatically typed `Astro.props`, plus it leads users to this pattern automatically by being a native API.

My personal suggestion would be to see if `props` couldn't be made lazy in Astro.
